janusgraph-hadoop-parent/janusgraph-hadoop-core/src/main/java/org/janusgraph/hadoop/formats/cassandra/CqlBridgeRecordReader.java

```diff
@@ -0,0 +1,372 @@
+import java.util.Map;
+
+import static java.util.stream.Collectors.toList;
+
+/**
+ * <p> Background: The {@link org.apache.cassandra.hadoop.cql3.CqlRecordReader} class has changed
+ * significantly in Cassandra-3 from Cassandra-2. This class acts as a bridge between
+ * CqlRecordReader in Cassandra-2 to Cassandra-3. In essence, this class recreates CqlRecordReader
+ * from Cassandra-3 without referring to it (because otherwise we'd get functionality from
+ * CqlRecordReader on Cassandra-2 and we don't want it). </p>
+ *
+ * @see <a href="https://github.com/JanusGraph/janusgraph/issues/172">Issue 172</a>.
+ */
+@Unstable // Because it should go away with JanusGraph upgrading to Cassandra-3 for OLAP
+@Deprecated // Because this class should already be on its path to deprecation
+public class CqlBridgeRecordReader extends RecordReader<StaticBuffer, Iterable<Entry>> {
+
+    /* Implementation note: This is inspired by Cassandra-3's org/apache/cassandra/hadoop/cql3/CqlRecordReader.java */
+    private static final Logger log = LoggerFactory.getLogger(CqlBridgeRecordReader.class);
+
+    private ColumnFamilySplit split;
+    private DistinctKeyIterator distinctKeyIterator;
+    private int totalRowCount; // total number of rows to fetch
+    private String keyspace;
+    private String cfName;
+    private String cqlQuery;
+    private Cluster cluster;
+    private Session session;
+    private IPartitioner partitioner;
+    private String inputColumns;
+    private String userDefinedWhereClauses;
+
+    private List<String> partitionKeys = new ArrayList<>();
+
+    // partition keys -- key aliases
+    private LinkedHashMap<String, Boolean> partitionBoundColumns = Maps.newLinkedHashMap();
+    private int nativeProtocolVersion = 1;
+
+    // binary type mapping code from CassandraBinaryRecordReader
+    private KV currentKV;
+
+    CqlBridgeRecordReader() { //package private
+        super();
+    }
+
+    @Override
+    public void initialize(InputSplit split, TaskAttemptContext context) throws IOException {
+        this.split = (ColumnFamilySplit) split;
+        Configuration conf = HadoopCompat.getConfiguration(context);
+        totalRowCount = (this.split.getLength() < Long.MAX_VALUE)
+            ? (int) this.split.getLength()
+            : ConfigHelper.getInputSplitSize(conf);
+        cfName = ConfigHelper.getInputColumnFamily(conf);
+        keyspace = ConfigHelper.getInputKeyspace(conf);
+        partitioner = ConfigHelper.getInputPartitioner(conf);
+        inputColumns = CqlConfigHelper.getInputcolumns(conf);
+        userDefinedWhereClauses = CqlConfigHelper.getInputWhereClauses(conf);
+
+        try {
+            if (cluster != null) {
+                return;
+            }
+            // create a Cluster instance
+            String[] locations = split.getLocations();
+//            cluster = CqlConfigHelper.getInputCluster(locations, conf);
+            // disregard the conf as it brings some unforeseen issues.
```

Comment:
Great news on the testing @kedarmhaswade. I combined two questions there, which was confusing. Thanks for the info on the `close` behavior. I think we'll need to provide our own config helper to at least pull out some of the basic cluster connection info so that we can support passing on settings related to auth, etc to the driver. Right now if I'm reading correctly, this won't work against a secured cluster, correct me if I'm wrong though.

I tested 2.1.9 more, and I'm running into the same Cassandra CPU issue across multiple OS'es so I think at this point, we should leave both record reader options in until we upgrade our bundled Cassandra. 

I think we're close to merging this PR, how about this to finish things up?

1. Address @mbrukman's comments related to the copyright header, namely pull that ASF header back in and then grab the contributor list from the Apache git repo and add that to our AUTHORS/CONTRIBUTORS and NOTICE file.
2. Update the code so that it uses a config parameter instead of a system property to enable this functionality the C* 3 support.

I think we can address improvement to the driver configuration options in a following PR.